app/src/server/tasks/evaluateTestSetEntries.task.ts

```diff
@@ -0,0 +1,428 @@
+      }
+    }
+
+    const [firstEntry, secondEntry] = entries;
+    const [firstEntryId, secondEntryId] = entryIds;
+
+    if (!firstEntry?.output || !secondEntry?.output || !firstEntryId || !secondEntryId) {
+      await prisma.datasetEvalResult.updateMany({
+        where: {
+          id: {
+            in: [firstResult.id, secondResult.id],
+          },
+        },
+        data: {
+          status: "ERROR",
+          errorMessage: "Error preparing for evaluation",
+        },
+      });
+      return;
+    }
+
+    let explanation;
+    let judgement;
+
+    const instructions = firstResult.datasetEvalDatasetEntry.datasetEval.instructions;
+
+    try {
+      const input: ChatCompletionCreateParams = {
+        model: "gpt-4",
+        messages: [
+          {
+            role: "system",
+            content: `You are an intelligent and fair judge of chatbots. Evaluate the following two chatbot responses and choose which one is better in the following way: ${
+              instructions ?? ""
+            }`,
+          },
+          {
+            role: "user",
+            content: formatDatasetEntryInputInstructions(
+              firstResult.datasetEvalDatasetEntry.datasetEntry,
+            ),
+          },
+          {
+            role: "user",
+            content: `This is what Alice said:\n\n${JSON.stringify(firstEntry.output)}`,
+          },
+          {
+            role: "user",
+            content: `This is what Bob said:\n\n${JSON.stringify(secondEntry.output)}`,
+          },
+          {
+            role: "user",
+            content: `Which response is better in the context of the task? Remember to pay attention to the following: ${
+              instructions ?? ""
+            }`,
+          },
+        ],
+        tools: [
+          {
+            type: "function",
+            function: {
+              name: "record_score",
+              parameters: {
+                type: "object",
+                properties: {
+                  explanation: {
+                    type: "string",
+                    description: "An explanation of why you chose the response you did",
+                  },
+                  judgement: {
+                    type: "string",
+                    enum: JUDGEMENT_OPTIONS,
+                    description: "Whose response is better in the context of the task?",
+                  },
+                },
+                required: ["explanation", "judgement"],
+              },
+            },
+          },
+        ],
+      };
+
+      const response = await getOpenaiCompletion(
+        firstResult.datasetEvalDatasetEntry.datasetEntry.dataset.projectId,
+        input,
+      );
+
+      const args = response.choices[0]?.message?.tool_calls?.[0]?.function?.arguments;
+
+      if (!args) throw new Error("No arguments returned" + JSON.stringify(response));
+
+      const parsedArgs = JSON.parse(args);
+
+      if (!parsedArgs["explanation"] || !parsedArgs["judgement"]) {
```

Comment:
Easier to just define a zod object for this and use that to validate. You can also use https://www.npmjs.com/package/zod-to-json-schema to use the same zod object as the function definition you pass to openai.

Reply:
Good idea, using zod now.